arborist/server.go

```diff
@@ -1304,13 +1304,43 @@ func (server *Server) handleUserRevokeAll(w http.ResponseWriter, r *http.Request
 func (server *Server) handleUserRevokePolicy(w http.ResponseWriter, r *http.Request) {
 	username := mux.Vars(r)["username"]
 	policyName := mux.Vars(r)["policyName"]
-	errResponse := revokeUserPolicy(server.db, username, policyName, getAuthZProvider(r))
-	if errResponse != nil {
-		errResponse.log.write(server.logger)
-		_ = errResponse.write(w, r)
-		return
+	authzProvider := getAuthZProvider(r)
+	policyInfo, _ := fetchPolicyInfo(server.db, username, policyName)
+
+	if policyInfo != nil {
+		server.logger.Info("Policy - {name: %s, authz_provider: %s, type: %s}",
+			policyInfo.Name, policyInfo.Authz_provider, policyInfo.Type)
+
+		if policyInfo.Type == "group" {
+			msg := fmt.Sprintf(
+				"Policy `%s` is assigned to the user through a group. Can not be revoked directly",
+				policyName)
+			errResponse := newErrorResponse(msg, http.StatusBadRequest, nil)
+			errResponse.log.write(server.logger)
+			_ = errResponse.write(w, r)
+			return
+		}
+
+		if policyInfo.Authz_provider == authzProvider.String {
```

Comment:
looking at the arborist code, i think you might need to check `authzProvider.Valid` first - if `Valid` is false, it means there is no authz provider, in which case you can revoke whatever

Reply:
If let us say, the `authz` field in the arborist db is `null` for a particular policy we are revoking and user has given a certain `authz` in the header. Then, do we `revoke` or `do nothing` in that case?

Reply:
neither, i would say the correct response is "unauthorized", but not 100% sure on that one, you might want to check the behavior of other endpoints in this case

Reply:
I didn't find an exact reference to another end point with this, but what you've mentioned made sense to me. Implemented the same.